Desktop integration

kermieisinthehouse opened the pull request:
## Desktop Integration

#### 💡 CI will not pass until a new compiler image with the `zip` package is pushed to dockerhub.

### ✨ System Tray Integration
(Windows / MacOS) The Stash server now lives in the system tray by default, instead of opening a shell window. This should be less off-putting to nontechnical people, and a little bit of a cleaner experience. This will detect when it should and shouldn't enable, and will keep old behavior if run from a terminal. `./stash` and  `>stash-win.exe` will continue to work exactly the same. The tray icon allows quick access to a few parts of the interface, and allows you to close the server. On the first run of this version, a notification will inform the user of the new location. Adding shortcuts is easy, future state can include a "Scan" button, etc.

<sub>This is not available on Linux because the relevant libraries do not support static linking, and we distribute static executables- it is otherwise functional when dynamically linked.</sub>

| Mac | Windows |
|:---:|:---:|
|![Screenshot from 2021-11-27 00-19-11](https://user-images.githubusercontent.com/84814418/143726361-04766602-8124-4d1a-aeff-94556fcac474.png)|![Screenshot from 2021-11-27 00-58-03](https://user-images.githubusercontent.com/84814418/143726364-92d85f64-d79d-4b69-83c9-87060515291c.png)|

### ✨ Notifications
(Windows / MacOS / Linux) A desktop notification system has been set up, for sending information to the user. There is an easy opt-out, and as of this patch, it will only notify the user when a task completes. In the future, this could be exposed to plugins / expanded.

![Screenshot from 2021-11-27 18-48-59](https://user-images.githubusercontent.com/84814418/143726947-98c2821c-8ef2-4ca1-8ec4-de724b16b6a4.png)

<img width="318" alt="Capture" src="https://user-images.githubusercontent.com/84814418/143727022-d6be7e43-6d41-4e8b-ac81-3c081aeb6ca8.PNG">

### ✨ MacOS Bundling
(MacOS) The build chain has been updated so that Mac executables are shipped as .app bundles, which is the idiomatic way of distributing applications. We gain (sort-of, we have signing issues) one-click launching, decorations, and access to more APIs by shipping like this. No more "chmod" for mac users, just download and run. There is now a single universal bundle for both MacOS architectures.
***Like many of my PRs, this requires a compiler image update: We are missing the `zip` package***

### ✨ Iconography
(Windows / MacOS) The executables are now decorated with an icon! There is a new script that allows updating every instance of the icon across the entire codebase from one master render. Supply it with a 1024x1024 png image to update the Stash logo programmatically in every place it appears. For now I've put on the glowing box concept, but this can be easily changed pending the results of #2049


| Mac | Windows |
|:---:|:---:|
|![Screenshot from 2021-11-27 18-19-07](https://user-images.githubusercontent.com/84814418/143726403-f7e38451-20ed-4f01-b768-c9e550cae40e.png)|![Screenshot from 2021-11-27 00-58-50](https://user-images.githubusercontent.com/84814418/143726410-f7f0bebb-c4d7-4de3-8720-c9a55c7b1421.png)

This PR has been tested in VMs of all of our supported platforms.

Resolves #594.

kermieisinthehouse commented:
Closes #594, #980, and forever stops issues like #1174

Some more details:
The tray menu during setup is just the Open/Close buttons, afterwards it uses the UI tabs that are enabled in settings.

Still TODO: 
Notifications do not feature an icon. Most of the APIs take a path to a file, so I'm thinking about (over-)writing an icon to the stash config folder on start, and passing a path to that. Perhaps using /tmp on platforms that have that.

I also want to add a channel to Config.go that alerts when a "large" change has been made, and that the systray has to re-render, e.g. the system setup is finalized, the UI tabs are changed, and (in the future) when an update is available.

kermieisinthehouse commented:
Notifications have icons now, and it passes the CI pipeline. I have kicked down the road systray updates, as this is buggy on windows and mac. I'll need to hack on the underlying library a little bit more to get that solved. Also, fixed an unrelated ffmpeg download bug on windows.

This is otherwise functional, systray works on mac / windows correctly. Notifications work on all platforms.

kermieisinthehouse commented:
Known issue: Windows does not have output when started from console. Exploring remediations for this

kermieisinthehouse commented:
All reported issues resolved.

- Windows shell behavior is now correct
- Console windows do not pop up when executing.
- Merged develop

kermieisinthehouse commented:
Thanks to @Belleyy for testing.

- Fixed a performance regression in windows ffmpeg (thanks @InfiniteTF )
- Fixed macos broken notifications
- Fixed logging http on windows

bnkai commented on pkg/api/server.go:
```diff
@@ -245,38 +249,32 @@ func Start(uiBox embed.FS, loginUIBox embed.FS) {
-			}
-		}
+	printVersion()
+	printLatestVersion(context.TODO())
```
this was run in a go function so that we dont need to wait for the response/timeout/network error and thus delay the start

kermieisinthehouse replied:
Fixed in latest

bnkai commented on pkg/desktop/desktop.go:
```diff
@@ -0,0 +1,156 @@
+		return false
+	}
+	// Check if stdin is a terminal
+	if term.IsTerminal(int(os.Stdin.Fd())) {
```
Just double checking for this one and linux systems
In the current dev this wasn't checked so stash could open a browser window when run from a terminal
With this applied only if you double click and execute stash from an explorer the browser window and the notifications are working

kermieisinthehouse replied:
That's intended. I got annoyed with Stash taking over my terminal with a `links` window when I was testing a build over ssh

bnkai commented on pkg/desktop/desktop.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func Shutdown() {
+	database.Close()
+	os.Exit(0)
```
shouldnt we check the error and log it and os.Exit(1)  if its != nil ?

kermieisinthehouse replied:
Fixed in latest

bnkai commented on pkg/manager/manager.go:
```diff
@@ -458,8 +427,12 @@ func (s *singleton) GetSystemStatus() *models.SystemStatus {
 
 // Shutdown gracefully stops the manager
-func (s *singleton) Shutdown() error {
+func (s *singleton) Shutdown(code int) {
```
Do we really need to force the exit code? 
If everything is ok it should be `0` otherwise `1`
In this case the only thing  we do is close the db, so if the err != nil we should then os.Exit(1) , not only log the error

bnkai replied:
**Not intended for the scope of this PR** 
It seems that the only way to reach the shutdown before this PR was for the user to either ctrl-c in the stash terminal or kill the stash process. With this PR this can be done  easily from the systray. We should find a proper way to shutdown  and wait for the manager (eg context cancelation) so that any running tasks are properly terminated before attempting to close the DB.

kermieisinthehouse replied:
Yeah, that's been on the to-do list a while.

I made it take an exit code so that a function that's bailing out can drop a code in. I can probably do an override if the parameter is 0 and it runs into an error closing the database, return a 1 instead.

bnkai replied:
Yeah that  sounds good.

kermieisinthehouse replied:
Added a check to see return error code if necessary

bnkai commented on pkg/api/authentication.go:
```diff
@@ -135,9 +135,4 @@ func securityActivateTripwireAccessedFromInternetWithoutAuth(c *config.Instance,
-	err = manager.GetInstance().Shutdown()
-	if err != nil {
-		logger.Error(err)
-	}
```
just double checking, shouldnt we shutdown here?

kermieisinthehouse replied:
No, it serves a short message about the misconfigured network. This feature is aimed toward people that don't read logs :)

bnkai commented:
Tests ok in linux (debian + xfce as a wm ) ( the notifications :-) )

bnkai commented on pkg/desktop/desktop_platform_linux.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func sendNotification(notificationTitle string, notificationText string) {
+	err := notificator.New(notificator.Options{
+		AppName:     "Stash",
+		DefaultIcon: getIconPath(),
+	}).Push(notificationTitle, notificationText, "", notificator.UR_NORMAL)
```
If its only for linux and only used here the included `notificator` library only does  the below and could be omitted
```golang
err := exec.Command("notify-send", "-i", getIconPath(), notificationTitle, notificationText, "-a", "Stash").Run()
```

kermieisinthehouse replied:
I bounced around the notification libraries a lot while testing, this might be true. I'll take a peek at this

bnkai replied:
It mentions Kdialog in the readme but it was removed in some point, only `notify-send` is used for linux

kermieisinthehouse replied:
It seems that's correct, I'll just call it directly then and save some vendor changes. I used that library for all three platforms but had to switch after the problems showed up, but it worked on Linux and it stayed.

kermieisinthehouse replied:
Removed the library, thanks for the tip on this

WithoutPants commented on Makefile:
```diff
@@ -46,9 +46,13 @@ ifndef OFFICIAL_BUILD
 endif
 
 build: pre-build
+ifdef IS_WIN
+PLATFORM_SPECIFIC_LDFLAGS := -H windowsgui
+endif
```
`IS_WIN` is intended for shell behaviours, not for detecting the windows platform. I had to manually include the windows flags in my local version because it didn't detect windows from my environment (which uses git bash).

kermieisinthehouse replied:
Fixed by adding a new variable for detecting the platform

WithoutPants commented on pkg/desktop/desktop.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// OpenURLInBrowser opens a browser to the Stash UI. Path can be an empty string for main page.
+func OpenURLInBrowser(force bool, path string) {
```
I don't think this needs to be public.

kermieisinthehouse replied:
Made private

WithoutPants commented on pkg/desktop/desktop.go:
```diff
@@ -0,0 +1,160 @@
+	// This can be done before actually starting the server, as modern browsers will
+	// automatically reload the page if a local port is closed at page load and then opened.
+	c := config.GetInstance()
+	if force || (!c.GetNoBrowser() && IsDesktop()) {
```
I think the `force` parameter should be removed and the other conditions should be moved into `Start` so that `OpenURLInBrowser` is only called under these conditions.

kermieisinthehouse replied:
Refactored to remove the force option and check before calling

WithoutPants commented on pkg/manager/config/config.go:
```diff
@@ -271,7 +282,22 @@ func (i *Instance) GetNoBrowser() bool {
+	return i.configUpdates
+}
+
+func (i *Instance) GetShowOneTimeMovedNotification() bool {
```
This needs explanation.

kermieisinthehouse replied:
Added a comment

WithoutPants commented on pkg/manager/config/config.go:
```diff
@@ -271,7 +282,22 @@ func (i *Instance) GetNoBrowser() bool {
+	return i.getBool(NotificationsEnabled)
+}
+
+func (i *Instance) GetConfigUpdatesChannel() chan int {
```
Is this used? The channel doesn't appear to be used at all.

kermieisinthehouse replied:
This is for updating the systray, which isn't working yet. It's a requirement for auto-update, so I've commented it out for now.

WithoutPants commented on pkg/desktop/desktop.go:
```diff
@@ -0,0 +1,160 @@
+	return true
+}
+
+func Shutdown() {
+	err := database.Close()
+	if err != nil {
+		logger.Errorf("Error closing database: %s", err)
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
```
This is (approximately) a duplicate of `manager.Shutdown` and I don't think it belongs here.

kermieisinthehouse replied:
This was here because of an import cycle between manager and desktop. I've moved shutdown() from manager to config to avoid the cycle, and the duplicated code.

WithoutPants reviewed (changes requested):
- running from shell does not output logger or http access messages to the shell (running via git bash)
- the favicon file should be accessible via the embedded `uiBox`. I don't understand why we're converting it into a go file.
- can we please use the existing stash icon until a decision is made to change it?
- there's a few commented out bits of code that either need to be removed or documented.

kermieisinthehouse commented:
>  running from shell does not output logger or http access messages to the shell (running via git bash)

Looking into this, the windows shell has been painful. I tested against cmd.exe and powershell, I'm not sure how git bash differs.
>    the favicon file should be accessible via the embedded `uiBox`. I don't understand why we're converting it into a go file.

That is still present and works as usual. It is a go file here so that notifications and the systray can be passed image data directly. Also, Windows and *nixes need the image in different formats. 

>    can we please use the existing stash icon until a decision is made to change it?

Changed. Again, the script included in this can swap it out super easily.

>   there's a few commented out bits of code that either need to be removed or documented.

I've added some more comments. Many of these will get used in the auto-update PR I'm working on, I started concurrently with this.

kermieisinthehouse commented:
@WithoutPants I'm having no issues with git bash, can you retest on the latest?

![Screenshot from 2022-01-07 17-03-17](https://user-images.githubusercontent.com/84814418/148625205-a624b4a5-7ec7-41b9-b0bf-a4e73f85acb5.png)

WithoutPants commented:
This bug appears to be limited to the git bash terminal within Visual Studio Code. Running stash via git bash externally logs as normal, as does running stash in cmd in Code.

WithoutPants commented:
I think there's also a potential issue with how this handles startup errors. Without any terminal to log to, if startup fails it just doesn't do anything at all. Probably not a show-stopper but it's a potential issue that we should try to resolve at some point.

WithoutPants commented on pkg/manager/config/init.go:
```diff
@@ -159,3 +161,17 @@ func makeOverrideConfig() *viper.Viper {
 	return viper
 }
+
+// Shutdown gracefully stops the application
+func Shutdown(code int) {
+	// TODO: Each part of the manager needs to gracefully stop at some point
+	// for now, we just close the database.
+	err := database.Close()
+	if err != nil {
+		logger.Errorf("Error closing database: %s", err)
+		if code == 0 {
+			os.Exit(1)
+		}
+	}
+	os.Exit(code)
+}
```
This does not belong in `config` - it belongs in `manager`.

To resolve the import cycle, I would suggest adding a new interface in `desktop`:

```
type ShutdownHandler interface {
	Shutdown(code int)
}
```

`desktop.Start` should be changed to accept a `ShutdownHandler` parameter. `manager` can create a small internal type that fulfils this interface by calling `Shutdown`.

WithoutPants replied:
I've implemented the suggested changes.

WithoutPants commented:
> > the favicon file should be accessible via the embedded `uiBox`. I don't understand why we're converting it into a go file.
> 
> That is still present and works as usual. It is a go file here so that notifications and the systray can be passed image data directly. Also, Windows and *nixes need the image in different formats.

Why can't this be passed the image data directly using the embedded data? It seems redundant.

WithoutPants commented:
> Why can't this be passed the image data directly using the embedded data? It seems redundant.

I've implemented the changes required for this, adding a png file to the ui directory. The generate script is amended so that it no longer generates go files. There is one pending issue with the script. If we're generating from some master png file, that png file should be committed somewhere in the code base, and the script should reference that file directly. I wasn't sure which original source file was used.

kermieisinthehouse commented:
Oh damn, thanks, I was only halfway done on my branch.

The current logo I used is https://github.com/kermieisinthehouse/stash-graphics/blob/main/stash-plain-box-padded-to-sqaure.png which I've pushed to this branch and re-ran the script.

Changes approved!
